// igniteui-angular/esm5/lib/services/csv/char-separated-value-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ExportUtilities } from '../exporter-common/export-utilities';
/**
 * @hidden
 */
var /**
 * @hidden
 */
CharSeparatedValueData = /** @class */ (function () {
    function CharSeparatedValueData(_data, valueDelimiter, _isTreeGridData) {
        if (_isTreeGridData === void 0) { _isTreeGridData = false; }
        this._data = _data;
        this._isTreeGridData = _isTreeGridData;
        this._headerRecord = '';
        this._dataRecords = '';
        this._eor = '\r\n';
        this._escapeCharacters = ['\r', '\n', '\r\n'];
        this._delimiterLength = 1;
        this._isSpecialData = false;
        this.setDelimiter(valueDelimiter);
    }
    /**
     * @return {?}
     */
    CharSeparatedValueData.prototype.prepareData = /**
     * @return {?}
     */
    function () {
        if (!this._data || this._data.length === 0) {
            return '';
        }
        this._data = this._data.map(function (item) { return item.rowData; });
        /** @type {?} */
        var keys = ExportUtilities.getKeysFromData(this._data);
        if (keys.length === 0) {
            return '';
        }
        this._isSpecialData = ExportUtilities.isSpecialData(this._data);
        this._escapeCharacters.push(this._delimiter);
        this._headerRecord = this.processHeaderRecord(keys, this._escapeCharacters);
        this._dataRecords = this.processDataRecords(this._data, keys, this._escapeCharacters);
        return this._headerRecord + this._dataRecords;
    };
    /**
     * @private
     * @param {?} value
     * @param {?} escapeChars
     * @return {?}
     */
    CharSeparatedValueData.prototype.processField = /**
     * @private
     * @param {?} value
     * @param {?} escapeChars
     * @return {?}
     */
    function (value, escapeChars) {
        /** @type {?} */
        var safeValue = ExportUtilities.hasValue(value) ? String(value) : '';
        if (escapeChars.some(function (v) { return safeValue.includes(v); })) {
            safeValue = "\"" + safeValue + "\"";
        }
        return safeValue + this._delimiter;
    };
    /**
     * @private
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    CharSeparatedValueData.prototype.processHeaderRecord = /**
     * @private
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    function (keys, escapeChars) {
        var e_1, _a;
        /** @type {?} */
        var recordData = '';
        try {
            for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                var keyName = keys_1_1.value;
                recordData += this.processField(keyName, this._escapeCharacters);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return recordData.slice(0, -this._delimiterLength) + this._eor;
    };
    /**
     * @private
     * @param {?} record
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    CharSeparatedValueData.prototype.processRecord = /**
     * @private
     * @param {?} record
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    function (record, keys, escapeChars) {
        var e_2, _a;
        /** @type {?} */
        var recordData = '';
        try {
            for (var keys_2 = tslib_1.__values(keys), keys_2_1 = keys_2.next(); !keys_2_1.done; keys_2_1 = keys_2.next()) {
                var keyName = keys_2_1.value;
                /** @type {?} */
                var value = (record[keyName] !== undefined) ? record[keyName] : this._isSpecialData ? record : '';
                recordData += this.processField(value, this._escapeCharacters);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (keys_2_1 && !keys_2_1.done && (_a = keys_2.return)) _a.call(keys_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return recordData.slice(0, -this._delimiterLength) + this._eor;
    };
    /**
     * @private
     * @param {?} currentData
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    CharSeparatedValueData.prototype.processDataRecords = /**
     * @private
     * @param {?} currentData
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    function (currentData, keys, escapeChars) {
        var e_3, _a;
        /** @type {?} */
        var dataRecords = '';
        try {
            for (var currentData_1 = tslib_1.__values(currentData), currentData_1_1 = currentData_1.next(); !currentData_1_1.done; currentData_1_1 = currentData_1.next()) {
                var row = currentData_1_1.value;
                dataRecords += this.processRecord(row, keys, escapeChars);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (currentData_1_1 && !currentData_1_1.done && (_a = currentData_1.return)) _a.call(currentData_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return dataRecords;
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    CharSeparatedValueData.prototype.setDelimiter = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._delimiter = value;
        this._delimiterLength = value.length;
    };
    return CharSeparatedValueData;
}());
/**
 * @hidden
 */
export { CharSeparatedValueData };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._headerRecord;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._dataRecords;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._eor;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._delimiter;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._escapeCharacters;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._delimiterLength;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._isSpecialData;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._data;
    /**
     * @type {?}
     * @private
     */
    CharSeparatedValueData.prototype._isTreeGridData;
}
