// igniteui-angular/esm2015/lib/grids/hierarchical-grid/selection.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { IgxSelectionAPIService } from '../../core/selection';
/**
 * @hidden
 */
export class IgxHierarchicalSelectionAPIService extends IgxSelectionAPIService {
    constructor() {
        super(...arguments);
        this.hSelection = new Map();
    }
    /**
     * @param {?} rootID
     * @param {?} parentID
     * @param {?} cell
     * @return {?}
     */
    add_sub_item(rootID, parentID, cell) {
        /** @type {?} */
        const selItem = new Map();
        selItem.set(parentID, cell);
        this.hSelection.set(rootID, selItem);
    }
    /**
     * @param {?} rootID
     * @return {?}
     */
    get_sub_item(rootID) {
        /** @type {?} */
        let selItem;
        /** @type {?} */
        const sel = this.hSelection.get(rootID);
        if (sel) {
            selItem = {
                gridID: this.hSelection.get(rootID).keys().next().value,
                cell: this.hSelection.get(rootID).values().next().value
            };
        }
        return selItem;
    }
    /**
     * @param {?} rootID
     * @return {?}
     */
    clear_sub_item(rootID) {
        return this.hSelection.set(rootID, null);
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    IgxHierarchicalSelectionAPIService.prototype.hSelection;
}
