// igniteui-angular/esm5/lib/animations/main.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IAnimationParams() { }
if (false) {
    /** @type {?} */
    IAnimationParams.prototype.delay;
    /** @type {?} */
    IAnimationParams.prototype.duration;
    /** @type {?} */
    IAnimationParams.prototype.easing;
    /** @type {?|undefined} */
    IAnimationParams.prototype.startOpacity;
    /** @type {?|undefined} */
    IAnimationParams.prototype.endOpacity;
    /** @type {?|undefined} */
    IAnimationParams.prototype.startAngle;
    /** @type {?|undefined} */
    IAnimationParams.prototype.endAngle;
    /** @type {?|undefined} */
    IAnimationParams.prototype.startDistance;
    /** @type {?|undefined} */
    IAnimationParams.prototype.endDistance;
    /** @type {?|undefined} */
    IAnimationParams.prototype.fromPosition;
    /** @type {?|undefined} */
    IAnimationParams.prototype.toPosition;
    /** @type {?|undefined} */
    IAnimationParams.prototype.fromScale;
    /** @type {?|undefined} */
    IAnimationParams.prototype.midScale;
    /** @type {?|undefined} */
    IAnimationParams.prototype.toScale;
    /** @type {?|undefined} */
    IAnimationParams.prototype.xPos;
    /** @type {?|undefined} */
    IAnimationParams.prototype.yPos;
    /** @type {?|undefined} */
    IAnimationParams.prototype.direction;
    /** @type {?|undefined} */
    IAnimationParams.prototype.rotateX;
    /** @type {?|undefined} */
    IAnimationParams.prototype.rotateY;
    /** @type {?|undefined} */
    IAnimationParams.prototype.rotateZ;
    /** @type {?|undefined} */
    IAnimationParams.prototype.startHeight;
    /** @type {?|undefined} */
    IAnimationParams.prototype.endHeight;
}
export { fadeIn, fadeOut } from './fade/index';
export { flipTop, flipRight, flipBottom, flipLeft, flipHorFwd, flipHorBck, flipVerFwd, flipVerBck } from './flip/index';
export { rotateInCenter, rotateInTop, rotateInRight, rotateInLeft, rotateInBottom, rotateInTr, rotateInBr, rotateInBl, rotateInTl, rotateInDiagonal1, rotateInDiagonal2, rotateInHor, rotateInVer, rotateOutCenter, rotateOutTop, rotateOutRight, rotateOutLeft, rotateOutBottom, rotateOutTr, rotateOutBr, rotateOutBl, rotateOutTl, rotateOutDiagonal1, rotateOutDiagonal2, rotateOutHor, rotateOutVer } from './rotate/index';
export { shakeHor, shakeVer, shakeTop, shakeBottom, shakeRight, shakeLeft, shakeCenter, shakeTr, shakeBr, shakeBl, shakeTl, heartbeat, pulsateFwd, pulsateBck, blink } from './misc/index';
export { scaleInTop, scaleInRight, scaleInBottom, scaleInLeft, scaleInCenter, scaleInTr, scaleInBr, scaleInBl, scaleInTl, scaleInVerTop, scaleInVerBottom, scaleInVerCenter, scaleInHorCenter, scaleInHorLeft, scaleInHorRight, scaleOutTop, scaleOutRight, scaleOutBottom, scaleOutLeft, scaleOutCenter, scaleOutTr, scaleOutBr, scaleOutBl, scaleOutTl, scaleOutVerTop, scaleOutVerBottom, scaleOutVerCenter, scaleOutHorCenter, scaleOutHorLeft, scaleOutHorRight } from './scale/index';
export { slideInTop, slideInRight, slideInBottom, slideInLeft, slideInTr, slideInBr, slideInBl, slideInTl, slideOutTop, slideOutBottom, slideOutRight, slideOutLeft, slideOutTr, slideOutBr, slideOutBl, slideOutTl } from './slide/index';
export { swingInTopFwd, swingInRightFwd, swingInLeftFwd, swingInBottomFwd, swingInTopBck, swingInRightBck, swingInBottomBck, swingInLeftBck, swingOutTopFwd, swingOutRightFwd, swingOutBottomFwd, swingOutLefttFwd, swingOutTopBck, swingOutRightBck, swingOutBottomBck, swingOutLeftBck } from './swing/index';
export { growVerIn, growVerOut } from './grow/index';
