// igniteui-angular/esm2015/lib/expansion-panel/expansion-panel.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IgxExpansionPanelBase() { }
if (false) {
    /** @type {?} */
    IgxExpansionPanelBase.prototype.id;
    /** @type {?} */
    IgxExpansionPanelBase.prototype.cssClass;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxExpansionPanelBase.prototype.headerId;
    /** @type {?} */
    IgxExpansionPanelBase.prototype.collapsed;
    /** @type {?} */
    IgxExpansionPanelBase.prototype.animationSettings;
    /** @type {?} */
    IgxExpansionPanelBase.prototype.onCollapsed;
    /** @type {?} */
    IgxExpansionPanelBase.prototype.onExpanded;
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelBase.prototype.collapse = function (evt) { };
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelBase.prototype.expand = function (evt) { };
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelBase.prototype.toggle = function (evt) { };
}
/**
 * @hidden
 * @type {?}
 */
export const IGX_EXPANSION_PANEL_COMPONENT = 'IgxExpansionPanelToken';
/**
 * @record
 */
export function IExpansionPanelEventArgs() { }
if (false) {
    /** @type {?} */
    IExpansionPanelEventArgs.prototype.event;
    /** @type {?} */
    IExpansionPanelEventArgs.prototype.panel;
}
