// igniteui-angular/esm5/lib/grids/hierarchical-grid/selection.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { IgxSelectionAPIService } from '../../core/selection';
/**
 * @hidden
 */
var /**
 * @hidden
 */
IgxHierarchicalSelectionAPIService = /** @class */ (function (_super) {
    tslib_1.__extends(IgxHierarchicalSelectionAPIService, _super);
    function IgxHierarchicalSelectionAPIService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.hSelection = new Map();
        return _this;
    }
    /**
     * @param {?} rootID
     * @param {?} parentID
     * @param {?} cell
     * @return {?}
     */
    IgxHierarchicalSelectionAPIService.prototype.add_sub_item = /**
     * @param {?} rootID
     * @param {?} parentID
     * @param {?} cell
     * @return {?}
     */
    function (rootID, parentID, cell) {
        /** @type {?} */
        var selItem = new Map();
        selItem.set(parentID, cell);
        this.hSelection.set(rootID, selItem);
    };
    /**
     * @param {?} rootID
     * @return {?}
     */
    IgxHierarchicalSelectionAPIService.prototype.get_sub_item = /**
     * @param {?} rootID
     * @return {?}
     */
    function (rootID) {
        /** @type {?} */
        var selItem;
        /** @type {?} */
        var sel = this.hSelection.get(rootID);
        if (sel) {
            selItem = {
                gridID: this.hSelection.get(rootID).keys().next().value,
                cell: this.hSelection.get(rootID).values().next().value
            };
        }
        return selItem;
    };
    /**
     * @param {?} rootID
     * @return {?}
     */
    IgxHierarchicalSelectionAPIService.prototype.clear_sub_item = /**
     * @param {?} rootID
     * @return {?}
     */
    function (rootID) {
        return this.hSelection.set(rootID, null);
    };
    return IgxHierarchicalSelectionAPIService;
}(IgxSelectionAPIService));
/**
 * @hidden
 */
export { IgxHierarchicalSelectionAPIService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    IgxHierarchicalSelectionAPIService.prototype.hSelection;
}
